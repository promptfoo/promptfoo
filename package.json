{
  "name": "promptfoo",
  "description": "LLM eval & testing toolkit",
  "author": "Ian Webster",
  "version": "0.61.0",
  "license": "MIT",
  "type": "commonjs",
  "repository": "promptfoo/promptfoo",
  "main": "dist/src/index.js",
  "exports": {
    ".": {
      "import": "./dist/src/index.js",
      "require": "./dist/src/index.js"
    }
  },
  "workspaces": [
    "src/web/nextui",
    "site"
  ],
  "types": "dist/src/index.d.ts",
  "typings": "dist/src/index.d.ts",
  "files": [
    "dist"
  ],
  "engines": {
    "node": ">=18"
  },
  "bin": {
    "promptfoo": "dist/src/main.js"
  },
  "scripts": {
    "bin": "dist/src/main.js",
    "tsc": "tsc",
    "local": "ts-node --cwdMode --transpileOnly src/main.ts",
    "local:web": "cd src/web/nextui && npm run dev",
    "install:nextui": "cd src/web/nextui && npm install",
    "build:clean": "rm -rf dist",
    "build:nextui": "cd src/web/nextui && npm run build && { [ -z \"$NEXT_PUBLIC_PROMPTFOO_BUILD_STANDALONE_SERVER\" ] && cp -r out/ ../../../dist/src/web/nextui; } || true",
    "build:watch": "tsc --watch",
    "build": "tsc && cp src/*.html dist/src && cp src/python/wrapper.py dist/src/python && rm -rf dist/drizzle && cp -r drizzle/ dist/drizzle && npm run build:nextui && chmod +x dist/src/main.js",
    "test": "jest",
    "test:watch": "jest --watch",
    "format": "prettier -w .",
    "format:check": "prettier --check .",
    "db:migrate": "npx tsx src/migrate.ts",
    "db:generate": "npx drizzle-kit generate:sqlite",
    "audit:fix": "npm audit fix && cd src/web/nextui && npm audit fix && cd ../../site && npm audit fix",
    "prepublishOnly": "npm run build:clean && npm run build"
  },
  "peerDependencies": {
    "@aws-sdk/client-bedrock-runtime": "^3.458.0",
    "@azure/identity": "^4.0.0",
    "@azure/openai-assistants": "^1.0.0-beta.5",
    "@ibm-generative-ai/node-sdk": "^2.0.6",
    "google-auth-library": "^9.7.0",
    "googleapis": "^134.0.0",
    "langfuse": "^3.7.0"
  },
  "devDependencies": {
    "@aws-sdk/client-bedrock-runtime": "^3.458.0",
    "@azure/identity": "^4.0.0",
    "@types/async": "^3.2.20",
    "@types/better-sqlite3": "^7.6.10",
    "@types/cache-manager": "^4.0.2",
    "@types/cache-manager-fs-hash": "^0.0.1",
    "@types/cli-progress": "^3.11.0",
    "@types/compression": "^1.7.2",
    "@types/cors": "^2.8.13",
    "@types/debounce": "^1.2.1",
    "@types/express": "^4.17.17",
    "@types/glob": "^8.1.0",
    "@types/jest": "^29.5.1",
    "@types/js-yaml": "^4.0.5",
    "@types/node-fetch": "^2.6.4",
    "@types/nunjucks": "^3.2.2",
    "@types/opener": "^1.4.0",
    "@types/semver": "^7.5.0",
    "@types/uuid": "^9.0.2",
    "babel-jest": "^29.5.0",
    "cloudflare": "^3.2.0",
    "check-dependency-version-consistency": "^4.1.0",
    "drizzle-kit": "^0.20.13",
    "jest": "^29.5.0",
    "jest-watch-typeahead": "^2.2.2",
    "next": "13.4.13",
    "prettier": "^2.8.8",
    "ts-jest": "^29.1.0",
    "ts-node": "^10.9.1",
    "typescript": "^5.1.6"
  },
  "dependencies": {
    "@anthropic-ai/sdk": "^0.20.1",
    "@apidevtools/json-schema-ref-parser": "^10.1.0",
    "ajv": "^8.14.0",
    "ajv-formats": "^2.1.1",
    "async": "^3.2.4",
    "better-sqlite3": "^10.1.0",
    "cache-manager": "^4.1.0",
    "cache-manager-fs-hash": "^1.0.0",
    "chalk": "^4.1.2",
    "chokidar": "^3.5.3",
    "cli-progress": "^3.12.0",
    "cli-table3": "^0.6.3",
    "commander": "^10.0.1",
    "compression": "^1.7.4",
    "cors": "^2.8.5",
    "csv-parse": "^5.5.3",
    "csv-stringify": "^6.3.2",
    "debounce": "^1.2.1",
    "dedent": "^1.5.3",
    "dotenv": "^16.4.5",
    "drizzle-orm": "^0.29.3",
    "express": "^4.18.2",
    "fast-deep-equal": "^3.1.3",
    "fastest-levenshtein": "^1.0.16",
    "glob": "^10.2.6",
    "js-yaml": "^4.1.0",
    "mathjs": "^12.4.1",
    "node-fetch": "^2.6.7",
    "nunjucks": "^3.2.4",
    "openai": "^4.38.5",
    "opener": "^1.5.2",
    "proxy-agent": "^6.3.1",
    "python-shell": "^5.0.0",
    "replicate": "^0.27.1",
    "rouge": "git+https://github.com/kenlimmj/rouge.git#f35111b599aca55f1d4dc1d4a3d15e28e7f7c55f",
    "semver": "^7.5.3",
    "socket.io": "^4.7.2",
    "tiny-invariant": "^1.3.1",
    "uuid": "^9.0.0",
    "winston": "^3.8.2"
  }
}
