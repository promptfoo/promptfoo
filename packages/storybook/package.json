{
  "name": "storybook",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc -b && vite build",
    "lint": "eslint .",
    "preview": "vite preview",
    "storybook": "storybook dev -p 6006",
    "build-storybook": "storybook build"
  },
  "dependencies": {
    "@storybook/global": "^5.0.0",
    "promptfoo-toolkit": "*",
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "devDependencies": {
    "@chromatic-com/storybook": "^4.0.1",
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.1",
    "@eslint/js": "^9.36.0",
    "@mui/icons-material": "^7.3.4",
    "@mui/material": "^7.3.4",
    "@mui/system": "^7.3.3",
    "@storybook/addon-a11y": "^9.1.10",
    "@storybook/addon-docs": "^9.1.10",
    "@storybook/addon-themes": "^9.1.10",
    "@storybook/addon-vitest": "^9.1.10",
    "@storybook/react-vite": "^9.1.10",
    "@types/node": "^24.7.2",
    "@types/react": "^18.3.26",
    "@types/react-dom": "^18.3.7",
    "@vitejs/plugin-react": "^5.0.4",
    "@vitest/browser": "^3.2.4",
    "@vitest/coverage-v8": "^3.2.4",
    "eslint": "^9.36.0",
    "eslint-plugin-react-hooks": "^5.2.0",
    "eslint-plugin-react-refresh": "^0.4.22",
    "eslint-plugin-storybook": "^9.1.10",
    "globals": "^16.4.0",
    "storybook": "^9.1.10",
    "typescript": "^5.9.3",
    "typescript-eslint": "^8.45.0",
    "vite": "^7.1.9",
    "vitest": "^3.2.4"
  }
}
