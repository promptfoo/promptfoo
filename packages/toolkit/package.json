{
  "name": "promptfoo-toolkit",
  "version": "1.0.0",
  "description": "A React component library for promptfoo",
  "private": true,
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "style": "dist/style.css",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "style": "./dist/style.css",
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs"
    },
    "./style.css": "./dist/style.css"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "vite build",
    "dev": "vite build --watch",
    "type-check": "tsc --noEmit",
    "test": "vitest",
    "test:ci": "vitest run",
    "clean": "rm -rf dist"
  },
  "keywords": [
    "react",
    "components",
    "promptfoo",
    "toolkit",
    "ui"
  ],
  "author": "promptfoo team",
  "license": "MIT",
  "peerDependencies": {
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.1",
    "@mui/icons-material": "^7.3.2",
    "@mui/material": "^7.3.2",
    "@mui/system": "^7.3.2",
    "@mui/x-charts": "^7.29.1",
    "@testing-library/dom": "^10.4.1",
    "@testing-library/jest-dom": "^6.8.0",
    "@testing-library/react": "^16.3.0",
    "@testing-library/user-event": "^14.6.1",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "react-hook-form": "^7.64.0"
  },
  "devDependencies": {
    "@types/react": "^18.2.0",
    "@types/react-dom": "^18.2.0",
    "@vitejs/plugin-react": "^4.0.0",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "typescript": "^5.9.2",
    "vite": "^4.4.0",
    "vite-plugin-dts": "^3.5.0",
    "vitest": "^3.2.4"
  }
}
