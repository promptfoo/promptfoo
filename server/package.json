{
  "dependencies": {
    "google-auth-library": "^10.3.0"
  }
}
