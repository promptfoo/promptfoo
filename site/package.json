{
  "name": "promptfoo-docs",
  "version": "0.0.1",
  "license": "MIT",
  "scripts": {
    "docusaurus": "docusaurus",
    "start": "docusaurus start",
    "build": "docusaurus build",
    "swizzle": "docusaurus swizzle",
    "deploy": "docusaurus deploy",
    "clear": "docusaurus clear",
    "serve": "docusaurus serve",
    "write-translations": "docusaurus write-translations",
    "write-heading-ids": "docusaurus write-heading-ids",
    "typecheck": "tsc",
    "format": "prettier --ignore-path .gitignore -w ."
  },
  "dependencies": {
    "@docusaurus/core": "^3.0.1",
    "@docusaurus/plugin-google-gtag": "^3.0.1",
    "@docusaurus/preset-classic": "^3.0.1",
    "@docusaurus/theme-search-algolia": "^3.0.1",
    "@mdx-js/react": "^3.0.0",
    "clsx": "^1.2.1",
    "docusaurus-plugin-image-zoom": "^1.0.1",
    "prism-react-renderer": "^2.3.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@docusaurus/module-type-aliases": "^3.0.1",
    "@tsconfig/docusaurus": "^1.0.7",
    "typescript": "^4.9.5"
  },
  "browserslist": {
    "production": [
      ">0.5%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "engines": {
    "node": ">=18"
  }
}
