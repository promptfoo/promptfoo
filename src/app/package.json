{
  "name": "app",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "tsc": "tsc",
    "build": "tsc -b && vite build",
    "build:watch": "tsc -b --watch",
    "lint": "eslint .",
    "preview": "vite preview",
    "test": "vitest",
    "test:ci": "vitest run"
  },
  "dependencies": {
    "@emotion/react": "^11.14.0",
    "@emotion/styled": "^11.14.0",
    "@mui/icons-material": "^6.1.10",
    "@mui/material": "^6.1.10",
    "@mui/system": "^6.1.10",
    "@mui/x-charts": "^7.23.1",
    "@tanstack/react-table": "^8.20.5",
    "chart.js": "^4.4.7",
    "csv-parse": "^5.6.0",
    "csv-stringify": "^6.5.2",
    "diff": "^7.0.0",
    "fuzzysearch": "^1.0.3",
    "html2canvas": "^1.4.1",
    "idb-keyval": "^6.2.1",
    "js-yaml": "^4.1.0",
    "jspdf": "^2.5.2",
    "prismjs": "^1.29.0",
    "react": "^18.3.1",
    "react-dom": "^18.3.1",
    "react-error-boundary": "^4.1.2",
    "react-markdown": "^9.0.1",
    "react-router-dom": "^6.28.0",
    "react-simple-code-editor": "^0.14.1",
    "react-syntax-highlighter": "^15.6.1",
    "recharts": "^2.14.1",
    "remark-gfm": "^4.0.0",
    "socket.io-client": "^4.8.1",
    "use-debounce": "^10.0.4",
    "zustand": "^5.0.2"
  },
  "devDependencies": {
    "@eslint/js": "^9.16.0",
    "@testing-library/dom": "^10.4.0",
    "@testing-library/jest-dom": "^6.6.3",
    "@testing-library/react": "^16.1.0",
    "@testing-library/user-event": "^14.5.2",
    "@types/diff": "^6.0.0",
    "@types/fuzzysearch": "^1.0.2",
    "@types/js-yaml": "^4.0.9",
    "@types/node": "^22.10.1",
    "@types/prismjs": "^1.26.5",
    "@types/react": "^18.3.13",
    "@types/react-dom": "^18.3.1",
    "@types/react-router-dom": "^5.3.3",
    "@types/react-syntax-highlighter": "^15.5.13",
    "@types/uuid": "^10.0.0",
    "@vitejs/plugin-react": "^4.3.4",
    "eslint": "^9.16.0",
    "eslint-plugin-react-hooks": "^5.1.0",
    "eslint-plugin-react-refresh": "^0.4.16",
    "globals": "^15.13.0",
    "jsdom": "^25.0.1",
    "typescript": "^5.7.2",
    "typescript-eslint": "^8.18.0",
    "vite": "^5.4.11",
    "vite-plugin-node-polyfills": "^0.22.0",
    "vitest": "^2.1.8"
  }
}
