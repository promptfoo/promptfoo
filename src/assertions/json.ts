import type { ValidateFunction } from 'ajv';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import yaml from 'js-yaml';
import { getEnvBool } from '../envars';
import type { AssertionParams, GradingResult } from '../types';
import invariant from '../util/invariant';
import { extractJsonObjects } from '../util/json';

let ajvInstance: Ajv | null = null;

export function resetAjv(): void {
  if (process.env.NODE_ENV !== 'test') {
    throw new Error('resetAjv can only be called in test environment');
  }
  ajvInstance = null;
}

export function getAjv(): Ajv {
  if (!ajvInstance) {
    const ajvOptions: ConstructorParameters<typeof Ajv>[0] = {
      strictSchema: !getEnvBool('PROMPTFOO_DISABLE_AJV_STRICT_MODE'),
    };
    ajvInstance = new Ajv(ajvOptions);
    addFormats(ajvInstance);
  }
  return ajvInstance;
}

export function handleIsJson({
  outputString,
  renderedValue,
  inverse,
  valueFromScript,
  assertion,
}: AssertionParams): GradingResult {
  let parsedJson;
  let pass;
  try {
    parsedJson = JSON.parse(outputString);
    pass = !inverse;
  } catch {
    pass = inverse;
  }

  if (pass && renderedValue) {
    let validate: ValidateFunction;
    if (typeof renderedValue === 'string') {
      if (renderedValue.startsWith('file://')) {
        // Reference the JSON schema from external file
        const schema = valueFromScript;
        invariant(schema, 'is-json references a file that does not export a JSON schema');
        validate = getAjv().compile(schema as object);
      } else {
        const scheme = yaml.load(renderedValue) as object;
        validate = getAjv().compile(scheme);
      }
    } else if (typeof renderedValue === 'object') {
      validate = getAjv().compile(renderedValue);
    } else {
      throw new Error('is-json assertion must have a string or object value');
    }
    pass = validate(parsedJson);
    if (!pass) {
      return {
        pass,
        score: 0,
        reason: `JSON does not conform to the provided schema. Errors: ${getAjv().errorsText(
          validate.errors,
        )}`,
        assertion,
      };
    }
  }

  return {
    pass,
    score: pass ? 1 : 0,
    reason: pass ? 'Assertion passed' : 'Expected output to be valid JSON',
    assertion,
  };
}

export function handleContainsJson({
  assertion,
  renderedValue,
  outputString,
  inverse,
  valueFromScript,
}: AssertionParams): GradingResult {
  let errorMessage = 'Expected output to contain valid JSON';
  const jsonObjects = extractJsonObjects(outputString);
  let pass = inverse ? jsonObjects.length === 0 : jsonObjects.length > 0;
  for (const jsonObject of jsonObjects) {
    if (renderedValue) {
      let validate: ValidateFunction;
      if (typeof renderedValue === 'string') {
        if (renderedValue.startsWith('file://')) {
          // Reference the JSON schema from external file
          const schema = valueFromScript;
          invariant(schema, 'contains-json references a file that does not export a JSON schema');
          validate = getAjv().compile(schema as object);
        } else {
          const scheme = yaml.load(renderedValue) as object;
          validate = getAjv().compile(scheme);
        }
      } else if (typeof renderedValue === 'object') {
        validate = getAjv().compile(renderedValue);
      } else {
        throw new Error('contains-json assertion must have a string or object value');
      }
      pass = validate(jsonObject);
      if (pass) {
        break;
      } else {
        errorMessage = `JSON does not conform to the provided schema. Errors: ${getAjv().errorsText(
          validate.errors,
        )}`;
      }
    }
  }
  return {
    pass,
    score: pass ? 1 : 0,
    reason: pass ? 'Assertion passed' : errorMessage,
    assertion,
  };
}
