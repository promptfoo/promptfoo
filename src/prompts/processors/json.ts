import * as fs from 'fs';
import { Prompt } from '../../types';

/**
 * Processes a JSON file to extract prompts.
 * This function reads a JSON file and converts it to a `Prompt` object.
 *
 * @param filePath - The path to the JSON file.
 * @param prompt - The raw prompt data, used for labeling.
 * @returns An array of one `Prompt` object.
 * @throws Will throw an error if the file cannot be read.
 */
export function processJsonFile(filePath: string, prompt: Partial<Prompt>): Prompt[] {
  const fileContents = fs.readFileSync(filePath, 'utf8');
  // NOTE: We do not validate if this is a valid JSON file.
  return [
    {
      raw: fileContents,
      label: prompt.label || `${filePath}: ${fileContents}`,
    },
  ];
}
