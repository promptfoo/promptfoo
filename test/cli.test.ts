import { spawn } from 'child_process';
import { version } from '../package.json';

const runCliCommand = (args: string[]) => {
  return new Promise<{ stdout: string; stderr: string }>((resolve, reject) => {
    try {
      const child = spawn('npm', ['run', 'local', '--', ...args], { shell: true });
      let stdout = '';
      let stderr = '';

      child.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      child.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      child.on('close', (code) => {
        if (code === 0) {
          resolve({ stdout, stderr });
        } else {
          reject(new Error(`Command failed with exit code ${code}\n${stderr}`));
        }
      });
    } catch (error) {
      reject(error);
    }
  });
};

describe('CLI Tool', () => {
  describe('--help and --version', () => {
    it('should display help menu', async () => {
      const { stdout } = await runCliCommand(['--help']);
      const subset = stdout.split('ts-node --cwdMode --transpileOnly src/main.ts --help')[1];
      expect(subset).toMatchSnapshot();
      expect(stdout).toContain('Usage: main [options] [command]');
    });

    it('should display version', async () => {
      const { stdout } = await runCliCommand(['--version']);
      expect(stdout).toMatch(/\d+\.\d+\.\d+/);
      expect(stdout).toContain(version);
    });
  });
});
