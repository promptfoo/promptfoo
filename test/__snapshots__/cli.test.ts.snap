// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI Tool --help and --version should display help menu 1`] = `
"

Usage: main [options] [command]

Options:
  --version                   Print version
  -h, --help                  display help for command

Commands:
  init [options] [directory]  Initialize project with dummy files
  view [options] [directory]  Start browser ui
  share [options]             Create a shareable URL of your most recent eval
  cache                       Manage cache
  feedback [message]          Send feedback to the promptfoo developers
  generate                    Generate synthetic data
  eval [options]              Evaluate prompts
  list                        List various resources
  show [options] <id>         Show details of a specific resource
  delete [options] <id>       Delete various resources
  import <file>               Import an eval record from a JSON file
  export [options] <evalId>   Export an eval record to a JSON file
  help [command]              display help for command
"
`;
